[ONNX] Add quantization support to more single output ops

justinchuby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #83008

#80039

- Implement quantization support for single output ops 
  - quantized::sigmoid
  - quantized::instance_norm
  - aten::reshape
  - aten::reshape_as
  - aten::sum
  - aten::mean
  - aten::prod
  - aten::t
  - aten::numpy_T
  - aten::expand
  - aten::expand_as
  - aten::embedding
  - aten::embedding_bag
  - aten::view
  - aten::select
  - aten::eq
  - aten::ne
  - aten::gt
  - aten::lt
  - aten::le
  - aten::ge
  - quantized::layer_norm
  - aten::elu
  - aten::selu
  - aten::maximum
  - aten::minimum
  - aten::amax
  - aten::amin
  - aten::hardtanh
  - aten::hardswish
  - quantized::group_norm
  - aten::as_strided
  - quantized::leaky_relu
  - aten::transpose
- Avoid modifying functions in `quantized_args` and have the wrapper closed over `scale` and `zero_point` instead (for purity)
- Remove magic number and assign it to INT64_MAX
- implement `_unpack_quantized_tensor` for handling quantized tensor unpacking to separate the logic from tuple unpacking and for clearer error handling

justinchuby committed:
[ONNX] Add quantization support to more ops 1/2

#80039

- Remove magic number and assign it to INT64_MAX

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Add quantization support to more ops 1/2"


#80039

- Remove magic number and assign it to INT64_MAX

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Add quantization support to more ops 1/2"


#80039

- Remove magic number and assign it to INT64_MAX

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Add quantization support to more ops 1/2"


#80039

- Remove magic number and assign it to INT64_MAX

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Add quantization support to more ops 1/2"


#80039

- Remove magic number and assign it to INT64_MAX

[ghstack-poisoned]

justinchuby commented:
Commit message:

- Remove magic number and assign it to INT64_MAX
- implement `_unpack_quantized_tensor` for handling quantized tensor unpacking to separate the logic from tuple unpacking
-

justinchuby committed:
Update on "[ONNX] Add quantization support to more single output ops"


#80039

- Remove magic number and assign it to INT64_MAX

[ghstack-poisoned]

BowenBao commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -11748,26 +11748,104 @@ def test_quantized_conv2d_relu(self):
+        class Model(torch.nn.Module):
+            def __init__(self, function):
+                super().__init__()
+                self.function = function
```
`function` is not accurate in most cases, where it is `module` or `sub_module`.

justinchuby replied:
Updated to function_or_module. What do you think?

BowenBao replied:
sgtm

BowenBao commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -11748,26 +11748,104 @@ def test_quantized_conv2d_relu(self):
+                torch.nn.ReLU(),
+                name="relu",
+            ),
+            # common_utils.subtest(
```
Is this still in progress? Perhaps better to either skip with reason or remove from this PR.

justinchuby replied:
It is a WIP for preliminary reviews. There are four tests I needed to fix

justinchuby replied:
Done

BowenBao commented on torch/onnx/symbolic_helper.py:
```diff
@@ -348,44 +369,55 @@ def q_foo(g, x, y):
             # TODO(justinchuby): Only single output is supported for now. We may want to
             # support multiple outputs in the future.
-            output = fn(g, *dequantized_args, **kwargs)
+            output = fn(g, *maybe_dequantized_args, **kwargs)
```
What is the rationale behind the renaming. Is empty list part of expected/accepted outcome?

justinchuby replied:
The thinking was that some ops were originally non-quantized and thus not "dequantized." Thoughts?

BowenBao replied:
Perhaps `unquantized_args` or `nonquantized_args` is better? The `maybe` word gave me a feeling that there are some quantized args that were unable to get dequantized. IIUC, the args should all become not quantized at this point, hence the mentioned names (or the original name) in the beginning of this comment.

justinchuby replied:
changed to non_quantized_args

justinchuby committed:
Update on "[ONNX] Add quantization support to more single output ops"


#80039

- Remove magic number and assign it to INT64_MAX

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Add quantization support to more single output ops"


#80039

- Remove magic number and assign it to INT64_MAX

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Add quantization support to more single output ops"


#80039

- Remove magic number and assign it to INT64_MAX

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Add quantization support to more single output ops"


#80039

- Implement quantization support for single output ops 
  - quantized::sigmoid
  - quantized::instance_norm
  - aten::reshape
  - aten::reshape_as
  - aten::sum
  - aten::mean
  - aten::prod
  - aten::t
  - aten::numpy_T
  - aten::expand
  - aten::expand_as
  - aten::embedding
  - aten::embedding_bag
  - aten::view
  - aten::select
  - aten::eq
  - aten::ne
  - aten::gt
  - aten::lt
  - aten::le
  - aten::ge
  - quantized::layer_norm
  - aten::elu
  - aten::selu
  - aten::maximum
  - aten::minimum
  - aten::amax
  - aten::amin
  - aten::hardtanh
  - aten::hardswish
  - quantized::group_norm
  - aten::as_strided
  - quantized::leaky_relu
  - aten::transpose
- Avoid modifying functions in `quantized_args` and have the wrapper closed over `scale` and `zero_point` instead (for purity)
- Remove magic number and assign it to INT64_MAX
- implement `_unpack_quantized_tensor` for handling quantized tensor unpacking to separate the logic from tuple unpacking and for clearer error handling

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Add quantization support to more single output ops"


#80039

- Implement quantization support for single output ops 
  - quantized::sigmoid
  - quantized::instance_norm
  - aten::reshape
  - aten::reshape_as
  - aten::sum
  - aten::mean
  - aten::prod
  - aten::t
  - aten::numpy_T
  - aten::expand
  - aten::expand_as
  - aten::embedding
  - aten::embedding_bag
  - aten::view
  - aten::select
  - aten::eq
  - aten::ne
  - aten::gt
  - aten::lt
  - aten::le
  - aten::ge
  - quantized::layer_norm
  - aten::elu
  - aten::selu
  - aten::maximum
  - aten::minimum
  - aten::amax
  - aten::amin
  - aten::hardtanh
  - aten::hardswish
  - quantized::group_norm
  - aten::as_strided
  - quantized::leaky_relu
  - aten::transpose
- Avoid modifying functions in `quantized_args` and have the wrapper closed over `scale` and `zero_point` instead (for purity)
- Remove magic number and assign it to INT64_MAX
- implement `_unpack_quantized_tensor` for handling quantized tensor unpacking to separate the logic from tuple unpacking and for clearer error handling

[ghstack-poisoned]

justinchuby commented:
@BowenBao ptal

BowenBao commented on torch/onnx/symbolic_opset10.py:
```diff
@@ -610,6 +610,82 @@ def hardswish(g, x, op_scale, op_zero_point):
+    ):
+        input, input_scale, _, _ = symbolic_helper.dequantize_helper(g, q_input)
+
+        # Duplicated from opset9.instance_norm
```
Why not call the function directly? If that doesn't work please add the reasoning into comment too.

justinchuby replied:
There wasn't a good reason. Updated to call the function.

BowenBao commented:
Please checkout CI failure

justinchuby committed:
Update on "[ONNX] Add quantization support to more single output ops"


#80039

- Implement quantization support for single output ops 
  - quantized::sigmoid
  - quantized::instance_norm
  - aten::reshape
  - aten::reshape_as
  - aten::sum
  - aten::mean
  - aten::prod
  - aten::t
  - aten::numpy_T
  - aten::expand
  - aten::expand_as
  - aten::embedding
  - aten::embedding_bag
  - aten::view
  - aten::select
  - aten::eq
  - aten::ne
  - aten::gt
  - aten::lt
  - aten::le
  - aten::ge
  - quantized::layer_norm
  - aten::elu
  - aten::selu
  - aten::maximum
  - aten::minimum
  - aten::amax
  - aten::amin
  - aten::hardtanh
  - aten::hardswish
  - quantized::group_norm
  - aten::as_strided
  - quantized::leaky_relu
  - aten::transpose
- Avoid modifying functions in `quantized_args` and have the wrapper closed over `scale` and `zero_point` instead (for purity)
- Remove magic number and assign it to INT64_MAX
- implement `_unpack_quantized_tensor` for handling quantized tensor unpacking to separate the logic from tuple unpacking and for clearer error handling

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Add quantization support to more single output ops"


#80039

- Implement quantization support for single output ops 
  - quantized::sigmoid
  - quantized::instance_norm
  - aten::reshape
  - aten::reshape_as
  - aten::sum
  - aten::mean
  - aten::prod
  - aten::t
  - aten::numpy_T
  - aten::expand
  - aten::expand_as
  - aten::embedding
  - aten::embedding_bag
  - aten::view
  - aten::select
  - aten::eq
  - aten::ne
  - aten::gt
  - aten::lt
  - aten::le
  - aten::ge
  - quantized::layer_norm
  - aten::elu
  - aten::selu
  - aten::maximum
  - aten::minimum
  - aten::amax
  - aten::amin
  - aten::hardtanh
  - aten::hardswish
  - quantized::group_norm
  - aten::as_strided
  - quantized::leaky_relu
  - aten::transpose
- Avoid modifying functions in `quantized_args` and have the wrapper closed over `scale` and `zero_point` instead (for purity)
- Remove magic number and assign it to INT64_MAX
- implement `_unpack_quantized_tensor` for handling quantized tensor unpacking to separate the logic from tuple unpacking and for clearer error handling

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Add quantization support to more single output ops"


#80039

- Implement quantization support for single output ops 
  - quantized::sigmoid
  - quantized::instance_norm
  - aten::reshape
  - aten::reshape_as
  - aten::sum
  - aten::mean
  - aten::prod
  - aten::t
  - aten::numpy_T
  - aten::expand
  - aten::expand_as
  - aten::embedding
  - aten::embedding_bag
  - aten::view
  - aten::select
  - aten::eq
  - aten::ne
  - aten::gt
  - aten::lt
  - aten::le
  - aten::ge
  - quantized::layer_norm
  - aten::elu
  - aten::selu
  - aten::maximum
  - aten::minimum
  - aten::amax
  - aten::amin
  - aten::hardtanh
  - aten::hardswish
  - quantized::group_norm
  - aten::as_strided
  - quantized::leaky_relu
  - aten::transpose
- Avoid modifying functions in `quantized_args` and have the wrapper closed over `scale` and `zero_point` instead (for purity)
- Remove magic number and assign it to INT64_MAX
- implement `_unpack_quantized_tensor` for handling quantized tensor unpacking to separate the logic from tuple unpacking and for clearer error handling

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Add quantization support to more single output ops"


#80039

- Implement quantization support for single output ops 
  - quantized::sigmoid
  - quantized::instance_norm
  - aten::reshape
  - aten::reshape_as
  - aten::sum
  - aten::mean
  - aten::prod
  - aten::t
  - aten::numpy_T
  - aten::expand
  - aten::expand_as
  - aten::embedding
  - aten::embedding_bag
  - aten::view
  - aten::select
  - aten::eq
  - aten::ne
  - aten::gt
  - aten::lt
  - aten::le
  - aten::ge
  - quantized::layer_norm
  - aten::elu
  - aten::selu
  - aten::maximum
  - aten::minimum
  - aten::amax
  - aten::amin
  - aten::hardtanh
  - aten::hardswish
  - quantized::group_norm
  - aten::as_strided
  - quantized::leaky_relu
  - aten::transpose
- Avoid modifying functions in `quantized_args` and have the wrapper closed over `scale` and `zero_point` instead (for purity)
- Remove magic number and assign it to INT64_MAX
- implement `_unpack_quantized_tensor` for handling quantized tensor unpacking to separate the logic from tuple unpacking and for clearer error handling

[ghstack-poisoned]

justinchuby committed:
Update on "[ONNX] Add quantization support to more single output ops"


#80039

- Implement quantization support for single output ops 
  - quantized::sigmoid
  - quantized::instance_norm
  - aten::reshape
  - aten::reshape_as
  - aten::sum
  - aten::mean
  - aten::prod
  - aten::t
  - aten::numpy_T
  - aten::expand
  - aten::expand_as
  - aten::embedding
  - aten::embedding_bag
  - aten::view
  - aten::select
  - aten::eq
  - aten::ne
  - aten::gt
  - aten::lt
  - aten::le
  - aten::ge
  - quantized::layer_norm
  - aten::elu
  - aten::selu
  - aten::maximum
  - aten::minimum
  - aten::amax
  - aten::amin
  - aten::hardtanh
  - aten::hardswish
  - quantized::group_norm
  - aten::as_strided
  - quantized::leaky_relu
  - aten::transpose
- Avoid modifying functions in `quantized_args` and have the wrapper closed over `scale` and `zero_point` instead (for purity)
- Remove magic number and assign it to INT64_MAX
- implement `_unpack_quantized_tensor` for handling quantized tensor unpacking to separate the logic from tuple unpacking and for clearer error handling

[ghstack-poisoned]

BowenBao reviewed (approved):
🚀

justinchuby commented:
@pytorchbot merge -g

justinchuby commented:
@pytorchbot merge -f "All checks passed"
